Source/Initialization/DivCleaner/ProjectionDivCleaner.cpp

```diff
@@ -0,0 +1,389 @@
+
+    for (int lev = 0; lev < m_levels; ++lev)
+    {
+        // Default BoxArray and DistributionMap for initializing the output MultiFab, m_mf_output.
+        const amrex::BoxArray& ba = warpx.boxArray(lev);
+        const amrex::DistributionMapping& dmap = warpx.DistributionMap(lev);
+
+        m_solution[lev].reset();
+        m_source[lev].reset();
+
+        const auto tag1 = amrex::MFInfo().SetTag("div_cleaner_solution");
+        m_solution[lev] = std::make_unique<MultiFab>(amrex::convert(ba, IntVect::TheCellVector()),
+            dmap, ncomps, ng, tag1);
+        const auto tag2 = amrex::MFInfo().SetTag("div_cleaner_source");
+        m_source[lev] = std::make_unique<MultiFab>(amrex::convert(ba, IntVect::TheCellVector()),
+            dmap, ncomps, ng, tag2);
+
+        m_solution[lev]->setVal(0.0, ng);
+        m_source[lev]->setVal(0.0, ng);
+    }
+
+    m_h_stencil_coefs_x.resize(1);
+    m_h_stencil_coefs_x[0] = 1._rt/cell_size[0];
+    m_h_stencil_coefs_y.resize(1);
+    m_h_stencil_coefs_y[0] = 1._rt/cell_size[1];
+    m_h_stencil_coefs_z.resize(1);
+    m_h_stencil_coefs_z[0] = 1._rt/cell_size[2];
+
+    m_stencil_coefs_x.resize(m_h_stencil_coefs_x.size());
+    m_stencil_coefs_y.resize(m_h_stencil_coefs_y.size());
+    m_stencil_coefs_z.resize(m_h_stencil_coefs_z.size());
+
+    amrex::Gpu::copyAsync(amrex::Gpu::hostToDevice,
+                          m_h_stencil_coefs_x.begin(), m_h_stencil_coefs_x.end(),
+                          m_stencil_coefs_x.begin());
+    amrex::Gpu::copyAsync(amrex::Gpu::hostToDevice,
+                          m_h_stencil_coefs_y.begin(), m_h_stencil_coefs_y.end(),
+                          m_stencil_coefs_y.begin());
+    amrex::Gpu::copyAsync(amrex::Gpu::hostToDevice,
+                          m_h_stencil_coefs_z.begin(), m_h_stencil_coefs_z.end(),
+                          m_stencil_coefs_z.begin());
+    amrex::Gpu::synchronize();
+}
+
+void
+ProjectionDivCleaner::solve ()
+{
+    // Get WarpX object
+    auto & warpx = WarpX::GetInstance();
+
+    const auto& ba = warpx.boxArray();
+    const auto& dmap = warpx.DistributionMap();
+    const auto& geom = warpx.Geom();
+
+    // Pull boundary conditions from WarpX class
+    // bogus values are overwritten.
+    amrex::Array<LinOpBCType,AMREX_SPACEDIM> lobc({AMREX_D_DECL(LinOpBCType::bogus,
+                                                                LinOpBCType::bogus,
+                                                                LinOpBCType::bogus)});
+    amrex::Array<LinOpBCType,AMREX_SPACEDIM> hibc({AMREX_D_DECL(LinOpBCType::bogus,
+                                                                LinOpBCType::bogus,
+                                                                LinOpBCType::bogus)});
+
+#ifdef WARPX_DIM_RZ
+    if (geom[0].ProbLo(0) == 0){
+        lobc[0] = LinOpBCType::Neumann;
+
+        // handle the r_max boundary explicitly
+        if (WarpX::field_boundary_hi[0] == FieldBoundaryType::PEC) {
```

Comment:
This line is only reached when `ProbLo(0) == 0`. What happens otherwise?

This whole section of code could be greatly simplified using a map.

Reply:
I will take a look at it. We should probably also refactor the BC logic in the static solvers the same way. I grabbed similar logic from there and modified.  I think a separate PR would appropriate to clean up the BC logic in several spots.

Reply:
I updated this to use a std::map hash table.